qt/widgets/common/src/FitScriptGeneratorPresenter.cpp

```diff
@@ -0,0 +1,180 @@
+
+namespace MantidQt {
+namespace MantidWidgets {
+
+FitScriptGeneratorPresenter::FitScriptGeneratorPresenter(
+    IFitScriptGeneratorView *view, IFitScriptGeneratorModel *model,
+    QStringList const &workspaceNames, double startX, double endX)
+    : m_warnings(), m_view(view), m_model(model) {
+  m_view->subscribePresenter(this);
+  setWorkspaces(workspaceNames, startX, endX);
+}
+
+FitScriptGeneratorPresenter::~FitScriptGeneratorPresenter() {}
+
+void FitScriptGeneratorPresenter::notifyPresenter(ViewEvent const &event) {
+  switch (event) {
+  case ViewEvent::RemoveClicked:
+    handleRemoveClicked();
+    return;
+  case ViewEvent::AddClicked:
+    handleAddWorkspaceClicked();
+    return;
+  case ViewEvent::StartXChanged:
+    handleStartXChanged();
+    return;
+  case ViewEvent::EndXChanged:
+    handleEndXChanged();
+    return;
+  }
+
+  throw std::runtime_error("Failed to notify the FitScriptGeneratorPresenter.");
+}
+
+void FitScriptGeneratorPresenter::openFitScriptGenerator() { m_view->show(); }
+
+void FitScriptGeneratorPresenter::handleRemoveClicked() {
+  for (auto const &index : m_view->selectedRows()) {
+    auto const workspaceName = m_view->workspaceName(index);
+    auto const workspaceIndex = m_view->workspaceIndex(index);
+
+    m_view->removeWorkspaceDomain(workspaceName, workspaceIndex);
+    m_model->removeWorkspaceDomain(workspaceName, workspaceIndex);
+  }
+}
+
+void FitScriptGeneratorPresenter::handleAddWorkspaceClicked() {
+  if (m_view->openAddWorkspaceDialog()) {
+    auto const workspaces = m_view->getDialogWorkspaces();
+    auto const workspaceIndices = m_view->getDialogWorkspaceIndices();
+
+    if (!workspaces.empty() && !workspaceIndices.empty())
+      addWorkspaces(workspaces, workspaceIndices);
+  }
+}
+
+void FitScriptGeneratorPresenter::handleStartXChanged() {
+  auto const selectedRows = m_view->selectedRows();
+  if (!selectedRows.empty()) {
+    auto const workspaceName = m_view->workspaceName(selectedRows[0]);
+    auto const workspaceIndex = m_view->workspaceIndex(selectedRows[0]);
+    auto const startX = m_view->startX(selectedRows[0]);
+
+    updateStartX(workspaceName, workspaceIndex, startX);
+  }
+}
+
+void FitScriptGeneratorPresenter::handleEndXChanged() {
+  auto const selectedRows = m_view->selectedRows();
+  if (!selectedRows.empty()) {
+    auto const workspaceName = m_view->workspaceName(selectedRows[0]);
+    auto const workspaceIndex = m_view->workspaceIndex(selectedRows[0]);
+    auto const endX = m_view->endX(selectedRows[0]);
+
+    updateEndX(workspaceName, workspaceIndex, endX);
+  }
+}
+
+void FitScriptGeneratorPresenter::setWorkspaces(
+    QStringList const &workspaceNames, double startX, double endX) {
+  for (auto const &workspaceName : workspaceNames)
+    addWorkspace(workspaceName.toStdString(), startX, endX);
+  checkForWarningMessages();
+}
+
+void FitScriptGeneratorPresenter::addWorkspaces(
+    std::vector<MatrixWorkspace_const_sptr> const &workspaces,
```

Comment:
dont need the const

Reply:
The const is still needed because it is a vector of `MatrixWorkspace_const_sptr`

Reply:
oh yes